From 0eb2cb96ef5c417f8c1567a77bfe6bdb836add0a Mon Sep 17 00:00:00 2001
From: LiangHui Chen <lianghui@codeaurora.org>
Date: Mon, 29 Jan 2018 16:57:05 +0800
Subject: [PATCH] ivi-controller: enable ivi share function

1. Change the buffer sharing handle frome name to fd.
2. Fix simple-ivi-share app block issue.
3. Support UBWC buffer sharing.

Signed-off-by: LiangHui Chen <lianghui@codeaurora.org>
---
 .../simple-ivi-share/CMakeLists.txt                |   4 +
 .../simple-ivi-share/src/simple-ivi-share.c        | 117 +++++++++++++---
 protocol/ivi-share.xml                             |  10 +-
 weston-ivi-shell/CMakeLists.txt                    |   5 +
 weston-ivi-shell/src/ivi-share-gbm.c               | 154 ++++++++++++++-------
 weston-ivi-shell/src/ivi-share.c                   |  12 +-
 weston-ivi-shell/src/ivi-share.h                   |   7 +-
 7 files changed, 231 insertions(+), 78 deletions(-)

diff --git a/ivi-layermanagement-examples/simple-ivi-share/CMakeLists.txt b/ivi-layermanagement-examples/simple-ivi-share/CMakeLists.txt
index 7401d08..e54a08b 100644
--- a/ivi-layermanagement-examples/simple-ivi-share/CMakeLists.txt
+++ b/ivi-layermanagement-examples/simple-ivi-share/CMakeLists.txt
@@ -34,6 +34,7 @@ pkg_check_modules(GLESv2 glesv2 REQUIRED)
 pkg_check_modules(EGL egl REQUIRED)
 pkg_check_modules(WAYLAND_EGL wayland-egl REQUIRED)
 pkg_check_modules(WAYLAND_CLIENT wayland-client REQUIRED)
+pkg_check_modules(GBM gbm REQUIRED)
 
 GET_TARGET_PROPERTY(IVI_EXTENSION_PROTOCOL_DIRS ivi-extension-protocol INCLUDE_DIRECTORIES)
 
@@ -43,12 +44,14 @@ include_directories(
     ${GLESv2_INCLUDE_DIRS}
     ${EGL_INCLUDE_DIRS}
     ${WAYLAND_CLIENT_INCLUDE_DIR}
+    ${GBM_INCLUDE_DIRS}
 )
 
 link_directories(
     ${GLESv2_LIBRARY_DIRS}
     ${EGL_LIBRARY_DIRS}
     ${WAYLAND_CLIENT_LIBRARY_DIRS}
+    ${GBM_LIBRARY_DIRS}
 )
 
 SET(LIBS
@@ -58,6 +61,7 @@ SET(LIBS
     ${WAYLAND_CLIENT_LIBRARIES}
     ${WAYLAND_EGL_LIBRARIES}
     ${EGL_LIBRARIES}
+    ${GBM_LIBRARIES}
     m
 )
 
diff --git a/ivi-layermanagement-examples/simple-ivi-share/src/simple-ivi-share.c b/ivi-layermanagement-examples/simple-ivi-share/src/simple-ivi-share.c
index 56d98d9..a2ace5c 100644
--- a/ivi-layermanagement-examples/simple-ivi-share/src/simple-ivi-share.c
+++ b/ivi-layermanagement-examples/simple-ivi-share/src/simple-ivi-share.c
@@ -22,19 +22,22 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <fcntl.h>
 #include <assert.h>
 #include <signal.h>
-
 #include <wayland-client.h>
 #include <wayland-egl.h>
-
 #include <EGL/egl.h>
 #include <EGL/eglext.h>
 #include <GLES2/gl2.h>
 #include <GLES2/gl2ext.h>
+#include <drm/drm_fourcc.h>
 
 #include "ivi-application-client-protocol.h"
 #include "ivi-share-client-protocol.h"
+#include "gbm_priv.h"
+
+#define DRM_DEV_NAME "/dev/dri/card0"
 
 struct window;
 
@@ -82,7 +85,8 @@ struct window {
         uint32_t    width;
         uint32_t    height;
         uint32_t    stride;
-        uint32_t    name;
+        uint32_t    format;
+        int32_t     fd;
         uint32_t    input_caps;
         EGLImageKHR eglimage;
     } share_buffer;
@@ -92,6 +96,9 @@ struct window {
 
 static int running = 1;
 
+static int fd = -1;
+static struct gbm_device *gbm = NULL;
+
 /*----------------------------------------------------------------------------*/
 
 static PFNEGLCREATEIMAGEKHRPROC            pfEglCreateImageKHR;
@@ -163,42 +170,91 @@ redraw(struct window *window, struct wl_callback *callback, uint32_t time)
     eglSwapBuffers(display->egl.egldisplay, window->eglsurface);
 }
 
+static uint32_t
+convert_to_drm_format(uint32_t format)
+{
+    uint32_t drm_format = 0;
+
+    switch (format)
+    {
+        case IVI_SHARE_SURFACE_FORMAT_ABGR8888:
+            drm_format = DRM_FORMAT_ABGR8888;
+            break;
+        case IVI_SHARE_SURFACE_FORMAT_ARGB8888:
+            drm_format = DRM_FORMAT_ARGB8888;
+            break;
+        case IVI_SHARE_SURFACE_FORMAT_XRGB8888:
+            drm_format = DRM_FORMAT_XRGB8888;
+            break;
+        default:
+            break;
+    }
+
+    return drm_format;
+}
+
 static void
 handle_share_surface_damage(void *data, struct ivi_share_surface *share_surface,
-                            uint32_t name)
+                            int32_t fd, int32_t meta_fd)
 {
     struct window *window = data;
     struct display *display = window->display;
-    EGLint img_attribs[] = {
-        EGL_WIDTH,                  window->share_buffer.width,
-        EGL_HEIGHT,                 window->share_buffer.height,
-        EGL_DRM_BUFFER_STRIDE_MESA, window->share_buffer.stride,
-        EGL_DRM_BUFFER_FORMAT_MESA, EGL_DRM_BUFFER_FORMAT_ARGB32_MESA,
-        EGL_NONE
+    struct gbm_bo *bo = NULL;
+
+    struct gbm_buf_info  buf_info;
+    EGLint attribs[] = {
+         EGL_WIDTH, 0,
+         EGL_HEIGHT, 0,
+         EGL_LINUX_DRM_FOURCC_EXT, 0,
+         EGL_DMA_BUF_PLANE0_FD_EXT, 0,
+         EGL_DMA_BUF_PLANE0_OFFSET_EXT, 0,
+         EGL_DMA_BUF_PLANE0_PITCH_EXT, 0,
+         EGL_NONE
     };
 
-    if (name == 0) {
+    if (fd < 0) {
+        return;
+    }
+
+    attribs[1]  = window->share_buffer.width;
+    attribs[3]  = window->share_buffer.height;
+    attribs[5]  = convert_to_drm_format(window->share_buffer.format);
+    attribs[7]  = fd;
+    attribs[11] = window->share_buffer.stride;
+
+    buf_info.fd          = fd;
+    buf_info.metadata_fd = meta_fd;
+    buf_info.width       = window->share_buffer.width;
+    buf_info.height      = window->share_buffer.height;
+    buf_info.format      = convert_to_drm_format(window->share_buffer.format);
+
+    bo = gbm_bo_import(gbm, GBM_BO_IMPORT_GBM_BUF_TYPE, &buf_info, GBM_BO_USE_RENDERING);
+    if (bo == NULL) {
+        fprintf(stderr,"create gbm bo failed!");
         return;
     }
 
-    if (window->share_buffer.name != name) {
+    if (window->share_buffer.fd != fd) {
         if (window->share_buffer.eglimage != NULL) {
             pfEglDestroyImageKHR(display->egl.egldisplay,
                                  window->share_buffer.eglimage);
         }
 
         window->share_buffer.eglimage =
-            pfEglCreateImageKHR(display->egl.egldisplay, EGL_NO_CONTEXT,
-                                EGL_DRM_BUFFER_MESA, (EGLClientBuffer)name,
-                                img_attribs);
+            eglCreateImageKHR(display->egl.egldisplay, EGL_NO_CONTEXT,
+                              EGL_LINUX_DMA_BUF_EXT, NULL,
+                              attribs);
+
         if (window->share_buffer.eglimage == NULL) {
             fprintf(stderr, "failed to create EGLImageKHR: 0x%04x\n",
                 eglGetError());
         }
 
         /* update */
-        window->share_buffer.name = name;
+        window->share_buffer.fd = fd;
     }
+
+    gbm_bo_destroy(bo);
 }
 
 static void
@@ -221,10 +277,11 @@ handle_share_surface_configure(void *data, struct ivi_share_surface *share_surfa
         window->geometry.height = height;
     }
 
-    window->share_buffer.name   = 0;
+    window->share_buffer.fd     = -1;
     window->share_buffer.width  = width;
     window->share_buffer.height = height;
     window->share_buffer.stride = stride;
+    window->share_buffer.format = format;
 }
 
 static void
@@ -350,8 +407,9 @@ init_gl(struct window *window)
 
     window->gl.pos     = 0;
     window->gl.tex_pos = 1;
-    glBindAttribLocation(program, window->gl.pos, "pos");
-    glBindAttribLocation(program, window->gl.tex_pos, "tex_pos");
+
+    glBindAttribLocation(program, window->gl.pos, "position");
+    glBindAttribLocation(program, window->gl.tex_pos, "texcoord");
     glLinkProgram(program);
 
     glEnableVertexAttribArray(window->gl.pos);
@@ -415,6 +473,8 @@ create_window(struct window *window)
         return -1;
     }
 
+    eglSwapInterval(display->egl.egldisplay, 0);
+
     if (init_gl(window) < 0) {
         return -1;
     }
@@ -636,6 +696,11 @@ destroy_window(struct window *window)
     if (window->surface)
         wl_surface_destroy(window->surface);
 
+    if (window->share_buffer.eglimage != NULL) {
+        pfEglDestroyImageKHR(window->display->egl.egldisplay,
+                             window->share_buffer.eglimage);
+    }
+
     eglTerminate(window->display->egl.egldisplay);
     eglReleaseThread();
 }
@@ -681,6 +746,18 @@ main(int argc, char **argv)
     window.geometry.width = 250;
     window.geometry.height = 250;
 
+    fd = open(DRM_DEV_NAME, O_RDWR);
+    if (fd < 0) {
+        fprintf(stderr, "Open drm device failed!\n");
+        return 1;
+    }
+    gbm = gbm_create_device(fd);
+    if (gbm == NULL) {
+        fprintf(stderr, "Create gbm device failed!\n");
+        close(fd);
+        return 1;
+    }
+
     if (create_display(&display) < 0) {
         return 1;
     }
@@ -711,6 +788,8 @@ main(int argc, char **argv)
     printf("exit simple-ivi-share\n");
     destroy_window(&window);
     destroy_display(&display);
+    gbm_device_destroy(gbm);
+    close(fd);
 
     return 0;
 }
diff --git a/protocol/ivi-share.xml b/protocol/ivi-share.xml
index 9be7113..729ebc6 100644
--- a/protocol/ivi-share.xml
+++ b/protocol/ivi-share.xml
@@ -59,9 +59,10 @@
 
         <enum name="format">
             <description summary="share surface format"/>
-            <entry name="argb8888" value="0"/>
-            <entry name="xrgb8888" value="1"/>
-            <entry name="unknown" value="2"/>
+            <entry name="abgr8888" value="0"/>
+            <entry name="argb8888" value="1"/>
+            <entry name="xrgb8888" value="2"/>
+            <entry name="unknown" value="3"/>
         </enum>
 
         <event name="damage">
@@ -69,7 +70,8 @@
               Notify if shared ivi_surface is updated in client.
             </description>
 
-            <arg name="name" type="uint"/>
+            <arg name="fd" type="fd"/>
+            <arg name="meta_fd" type="fd"/>
         </event>
 
         <event name="configure">
diff --git a/weston-ivi-shell/CMakeLists.txt b/weston-ivi-shell/CMakeLists.txt
index d88694e..f4ff63d 100644
--- a/weston-ivi-shell/CMakeLists.txt
+++ b/weston-ivi-shell/CMakeLists.txt
@@ -31,6 +31,7 @@ find_package(Threads REQUIRED)
 if (IVI_SHARE)
     pkg_check_modules(GBM gbm REQUIRED)
     pkg_check_modules(LIBDRM libdrm REQUIRED)
+    pkg_check_modules(LIBEGL egl REQUIRED)
     ADD_DEFINITIONS("-DIVI_SHARE_ENABLE")
     SET(BUFFER_SHARING_SRC_FILES
         src/ivi-share.c
@@ -79,23 +80,27 @@ if (IVI_SHARE)
          ${include_directories}
          ${GBM_INCLUDE_DIRS}
          ${LIBDRM_INCLUDE_DIRS}
+         ${LIBEGL_INCLUDE_DIRS}
      )
 
      link_directories(
          ${link_directories}
          ${GBM_LIBRARY_DIRS}
          ${LIBDRM_LIBRARY_DIRS}
+         ${LIBEGL_LIBRARY_DIRS}
      )
 
      add_dependencies(${PROJECT_NAME}
          ${GBM_LIBRARIES}
          ${LIBDRM_LIBRARIES}
+         ${LIBEGL_LIBRARIES}
      )
 
      set(LIBS
          ${LIBS}
          ${GBM_LIBRARIES}
          ${LIBDRM_LIBRARIES}
+         ${LIBEGL_LIBRARIES}
      )
 endif (IVI_SHARE)
 
diff --git a/weston-ivi-shell/src/ivi-share-gbm.c b/weston-ivi-shell/src/ivi-share-gbm.c
index 310270b..8e56af8 100644
--- a/weston-ivi-shell/src/ivi-share-gbm.c
+++ b/weston-ivi-shell/src/ivi-share-gbm.c
@@ -24,18 +24,29 @@
 #include <string.h>
 #include <assert.h>
 #include <sys/time.h>
+#include <sys/mman.h>
 #include <xf86drm.h>
 #include <xf86drmMode.h>
 #include <gbm.h>
+#include <gbm_priv.h>
 #include <libudev.h>
 
+#include <EGL/egl.h>
+#include <EGL/eglext.h>
+#include <EGL/eglwaylandext.h>
+
 #include "wayland-server.h"
 #include "weston/compositor.h"
 #include "ivi-share-server-protocol.h"
 #include "ivi-controller-interface.h"
 #include "ivi-share.h"
 
-static uint32_t nativesurface_name;
+#define ALIGN(x, align) (((x) + ((align)-1)) & ~((align)-1))
+
+static int32_t nativesurface_fd = -1;
+static int32_t nativesurface_meta_fd = -1;
+
+static PFNEGLQUERYWAYLANDBUFFERWL query_wayland_buffer = NULL;
 
 /* copied from libinput-seat.h of weston-1.9.0 */
 struct udev_input {
@@ -93,77 +104,126 @@ struct drm_backend {
 	int32_t cursor_height;
 };
 
-uint32_t
-get_buffer_name(struct weston_surface *surface,
+
+
+int32_t
+get_buffer_fd(struct weston_surface *surface,
+                struct ivi_shell_share_ext *shell_ext)
+{
+     (void)surface;
+     return nativesurface_fd;
+}
+
+int32_t
+get_meta_buffer_fd(struct weston_surface *surface,
                 struct ivi_shell_share_ext *shell_ext)
 {
      (void)surface;
-     return nativesurface_name;
+     return nativesurface_meta_fd;
+}
+
+static uint32_t
+convert_to_ivi_share_format(uint32_t format)
+{
+    uint32_t ivi_share_format = IVI_SHARE_SURFACE_FORMAT_UNKNOWN;
+
+    switch (format)
+    {
+        case GBM_FORMAT_ABGR8888:
+            ivi_share_format = IVI_SHARE_SURFACE_FORMAT_ABGR8888;
+            break;
+        case GBM_FORMAT_ARGB8888:
+            ivi_share_format = IVI_SHARE_SURFACE_FORMAT_ARGB8888;
+            break;
+        case GBM_FORMAT_XRGB8888:
+            ivi_share_format = IVI_SHARE_SURFACE_FORMAT_XRGB8888;
+            break;
+        default:
+            break;
+    }
+
+    return ivi_share_format;
 }
 
 uint32_t
 update_buffer_nativesurface(struct ivi_share_nativesurface *p_nativesurface,
                             struct ivi_shell_share_ext *shell_ext)
 {
+    struct drm_backend *backend = NULL;
+    struct linux_dmabuf_buffer *dmabuf = NULL;
+    struct weston_buffer *buffer = NULL;
+    struct gbm_buffer *gbm_buffer = NULL;
+    struct gbm_buf_info *buf_info = NULL;
+    struct wl_shm_buffer *shm_buffer = NULL;
+    EGLint format;
+    unsigned int alignment = 32;
+
     if (NULL == p_nativesurface || NULL == p_nativesurface->surface) {
         return IVI_SHAREBUFFER_NOT_AVAILABLE;
     }
 
-    struct drm_backend *backend =
-        (struct drm_backend*)p_nativesurface->surface->compositor->backend;
+    backend = (struct drm_backend*)p_nativesurface->surface->compositor->backend;
     if (NULL == backend) {
         return IVI_SHAREBUFFER_NOT_AVAILABLE;
     }
 
-    struct weston_buffer *buffer = p_nativesurface->surface->buffer_ref.buffer;
+    if (query_wayland_buffer == NULL)
+        query_wayland_buffer = (void *) eglGetProcAddress("eglQueryWaylandBufferWL");
+
+    buffer = p_nativesurface->surface->buffer_ref.buffer;
     if (!buffer) {
         return IVI_SHAREBUFFER_NOT_AVAILABLE;
     }
 
-    struct gbm_bo *bo = gbm_bo_import(backend->gbm, GBM_BO_IMPORT_WL_BUFFER,
-                                      buffer->legacy_buffer, GBM_BO_USE_SCANOUT);
-    if (!bo) {
-        weston_log("failed to import gbm_bo\n");
-        return IVI_SHAREBUFFER_INVALID;
-    }
+    shm_buffer = wl_shm_buffer_get(buffer->resource);
 
-    struct drm_gem_flink flink = {0};
-    flink.handle = gbm_bo_get_handle(bo).u32;
-    if (drmIoctl(gbm_device_get_fd(backend->gbm), DRM_IOCTL_GEM_FLINK, &flink) != 0) {
-        weston_log("gem_flink: returned non-zero failed\n");
-        gbm_bo_destroy(bo);
-        return IVI_SHAREBUFFER_INVALID;
+    if (shm_buffer) {
+        return IVI_SHAREBUFFER_INVALID; //don't support yet.
     }
-
-    uint32_t name   = flink.name;
-    uint32_t width  = gbm_bo_get_width(bo);
-    uint32_t height = gbm_bo_get_height(bo);
-    uint32_t stride = gbm_bo_get_stride(bo);
-    uint32_t format = IVI_SHARE_SURFACE_FORMAT_ARGB8888;
-    uint32_t ret    = IVI_SHAREBUFFER_STABLE;
-
-    nativesurface_name = name;
-
-    if (name != p_nativesurface->name) {
-        ret |= IVI_SHAREBUFFER_DAMAGE;
+    else if ((query_wayland_buffer != NULL) &&
+               query_wayland_buffer(eglGetCurrentDisplay(), (void *) buffer->resource,
+                                    EGL_TEXTURE_FORMAT, &format)) {
+        buf_info = wl_resource_get_user_data(buffer->resource);
+        int32_t  fd          = buf_info->fd;
+        int32_t  metadata_fd = buf_info->metadata_fd;
+        uint32_t width       = buf_info->width;
+        uint32_t height      = buf_info->height;
+        uint32_t stride      = ALIGN(width, alignment) * 4; //in bytes
+        uint32_t format      = convert_to_ivi_share_format(buf_info->format);
+        uint32_t ret         = IVI_SHAREBUFFER_STABLE;
+
+        nativesurface_fd = fd;
+        nativesurface_meta_fd = metadata_fd;
+
+        if (fd != p_nativesurface->fd) {
+            ret |= IVI_SHAREBUFFER_DAMAGE;
+        }
+        if (width != p_nativesurface->width) {
+            ret |= IVI_SHAREBUFFER_CONFIGURE;
+        }
+        if (height != p_nativesurface->height) {
+            ret |= IVI_SHAREBUFFER_CONFIGURE;
+        }
+        if (stride != p_nativesurface->stride) {
+            ret |= IVI_SHAREBUFFER_CONFIGURE;
+        }
+
+        p_nativesurface->fd     = fd;
+        p_nativesurface->width  = width;
+        p_nativesurface->height = height;
+        p_nativesurface->stride = stride;
+        p_nativesurface->format = format;
+
+        return ret;
     }
-    if (width != p_nativesurface->width) {
-        ret |= IVI_SHAREBUFFER_CONFIGURE;
+    else if (dmabuf = linux_dmabuf_buffer_get(buffer->resource)) {
+        return IVI_SHAREBUFFER_INVALID; //don't support yet.
     }
-    if (height != p_nativesurface->height) {
-        ret |= IVI_SHAREBUFFER_CONFIGURE;
+    else if (gbm_buffer = gbm_buffer_get(buffer->resource)) {
+        return IVI_SHAREBUFFER_INVALID; //don't support yet.
     }
-    if (stride != p_nativesurface->stride) {
-        ret |= IVI_SHAREBUFFER_CONFIGURE;
+    else {
+        return IVI_SHAREBUFFER_NOT_AVAILABLE;
     }
-
-    p_nativesurface->name   = name;
-    p_nativesurface->width  = width;
-    p_nativesurface->height = height;
-    p_nativesurface->stride = stride;
-    p_nativesurface->format = format;
-
-    gbm_bo_destroy(bo);
-
-    return ret;
 }
+
diff --git a/weston-ivi-shell/src/ivi-share.c b/weston-ivi-shell/src/ivi-share.c
index dedbf45..fd7530e 100644
--- a/weston-ivi-shell/src/ivi-share.c
+++ b/weston-ivi-shell/src/ivi-share.c
@@ -24,7 +24,6 @@
 
 #include "ivi-share.h"
 #include "ivi-share-server-protocol.h"
-
 #ifndef container_of
 #define container_of(ptr, type, member) ({                              \
         const __typeof__( ((type *)0)->member ) *__mptr = (ptr);        \
@@ -344,7 +343,7 @@ alloc_share_nativesurface(struct weston_surface *surface, uint32_t id, uint32_t
     nativesurface->id = id;
     nativesurface->surface = surface;
     nativesurface->bufferType = bufferType;
-    nativesurface->name = 0;
+    nativesurface->fd = -1;
     nativesurface->width = 0;
     nativesurface->height = 0;
     nativesurface->stride = 0;
@@ -551,17 +550,17 @@ send_configure(struct wl_resource *p_resource, uint32_t id,
                                      p_nativesurface->bufferType,
                                      p_nativesurface->width,
                                      p_nativesurface->height,
-                                     p_nativesurface->stride / 4,
+                                     p_nativesurface->stride,
                                      p_nativesurface->format);
 }
 
 static void
-send_damage(struct wl_resource *p_resource, uint32_t id, uint32_t name)
+send_damage(struct wl_resource *p_resource, uint32_t id, int32_t fd, int32_t meta_fd)
 {
     if (!p_resource)
         return;
 
-    ivi_share_surface_send_damage(p_resource, name);
+    ivi_share_surface_send_damage(p_resource, fd, meta_fd);
 }
 
 static void
@@ -611,7 +610,8 @@ send_to_client(struct ivi_share_nativesurface *p_nativesurface, uint32_t send_fl
         }
         if ((IVI_SHAREBUFFER_DAMAGE & send_flag) == IVI_SHAREBUFFER_DAMAGE) {
             send_damage(p_link->resource, p_nativesurface->id,
-                        get_buffer_name(p_nativesurface->surface, shell_ext));
+                        get_buffer_fd(p_nativesurface->surface, shell_ext),
+                        get_meta_buffer_fd(p_nativesurface->surface, shell_ext));
         }
     }
 }
diff --git a/weston-ivi-shell/src/ivi-share.h b/weston-ivi-shell/src/ivi-share.h
index 63dbd60..fd82f76 100644
--- a/weston-ivi-shell/src/ivi-share.h
+++ b/weston-ivi-shell/src/ivi-share.h
@@ -36,7 +36,7 @@ struct ivi_share_nativesurface
     struct wl_list link;            /* link                                       */
     struct wl_list client_list;     /* ivi_nativesurface_client_link list         */
     uint32_t bufferType;            /* buffer type (GBM only)                     */
-    uint32_t name;                  /* buffer name                                */
+    int32_t  fd;                    /* buffer fd                                  */
     uint32_t width;                 /* buffer width                               */
     uint32_t height;                /* buffer height                              */
     uint32_t stride;                /* buffer stride[LSB:byte]                    */
@@ -70,7 +70,10 @@ enum ivi_sharebuffer_updatetype
 int32_t setup_buffer_sharing(struct weston_compositor *wc,
                              const struct ivi_controller_interface *interface);
 
-uint32_t get_buffer_name(struct weston_surface *surface,
+int32_t get_buffer_fd(struct weston_surface *surface,
                          struct ivi_shell_share_ext *shell_ext);
+int32_t get_meta_buffer_fd(struct weston_surface *surface,
+                         struct ivi_shell_share_ext *shell_ext);
+
 uint32_t update_buffer_nativesurface(struct ivi_share_nativesurface *nativesurface,
                                      struct ivi_shell_share_ext *shell_ext);
-- 
1.9.1

